### PR TITLE
#285: #apply yields versions to be applied

### PR DESCRIPTION
**Pull request by ledestin**
https://github.com/zold-io/zold/issues/326#issuecomment-403288342

As requested, I `yield` the versions of upgrade scripts to run, from `VersionFile#apply`.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#367](https://github.com/zold-io/zold/issues/367) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#367](https://github.com/zold-io/zold/issues/367) is assigned to @botanicus[/z](https://www.0crat.com/u/botanicus), here is [why](https://www.0crat.com/footprint/CAZPZR9FS/1901047280); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: lib/zold/version_file.rb [hunk 1]
```diff
@@ -9,11 +9,22 @@ def initialize(path, log: Log::Verbose.new)
...
+
+      yield version
+    end
+
+    private
+
+    def current_data_version
+      return Gem::Version.new('0') unless File.exist?(version_file)
+
+      @current_data_version ||= Gem::Version.new(File.read(version_file))
```
Comment (botanicus) [inline lib/zold/version_file.rb]: @ledestin this will have "\n" in it. Are you sure `Gem::Version` Is OK with it? I'd be rather surprised.

**Comment by botanicus** [comment]
@ledestin it looks good, but the tests are missing and I think a new puzzle should be added, what's the next step?

File: lib/zold/version_file.rb (hunk 1 above)
Comment (ledestin) [reply lib/zold/version_file.rb]: @botanicus thanks, updated

**Comment by ledestin** [comment]
@botanicus the skipped upgrade manager tests I've added might be enough https://github.com/zold-io/zold/blob/master/test/test_upgrades.rb#L34

**Comment by ledestin** [comment]
@botanicus which puzzle do you mean?

**Comment by botanicus** [comment]
@ledestin well, I'm hardly an expert on PDD, but my limited understanding is that you do something small and add a puzzle to say what needs to be done next, so it's a string of things that needs to get done and the string doesn't get interrupted until the feature is fully implemented.

If my interpretation is +/- correct, than I'd expect a puzzle saying what the next step is.

**Comment by ledestin** [comment]
@botanicus done

**Comment by botanicus** [comment]
@ledestin looks good, could you please just resolve the merge conflict? I don't think rultor would be able to merge otherwise.

@yegor256 looks good to me, once the merge conflict that GH is showing is solved.

File: lib/zold/version_file.rb [hunk 2]
```diff
@@ -9,11 +9,26 @@ def initialize(path, log: Log::Verbose.new)
       @log = log
     end
 
-    # @todo #285:30min Replace this stub with functionality.
-    #  We need to run the script (`yield`) if the version of
-    #  the script is between the saved version and the current one.
     def apply(version)
-      @log.info("Version #{version} doesn't need to be applied.")
+      return unless Gem::Version.new(version) > current_data_version
+
```
Comment (yegor256) [inline lib/zold/version_file.rb]: @ledestin keep this in mind: https://www.yegor256.com/2014/11/03/empty-line-code-smell.html

File: lib/zold/version_file.rb [hunk 3]
```diff
@@ -9,11 +9,26 @@ def initialize(path, log: Log::Verbose.new)
...
+      yield version
+    end
+
+    private
+
+    def current_data_version
+      return Gem::Version.new('0') unless File.exist?(version_file)
+
```
Comment (yegor256) [inline lib/zold/version_file.rb]: @ledestin again here: https://www.yegor256.com/2014/11/03/empty-line-code-smell.html An empty line inside method body is an obvious sign of a problem in the method design. Never use them.

**Comment by yegor256** [comment]
@ledestin also, see my comments above. Also, I see a problem in method naming. Why one of them is `current_version` while another one is `read_version`? As a rule of thumb, use verbs for modifiers and nouns for readers.

**Comment by ledestin** [comment]
@botanicus please check

**Comment by botanicus** [comment]
@ledestin thanks, that's good for me.

**Comment by ledestin** [comment]
@yegor256 updated

**Comment by 0crat** [comment]
@botanicus[/z](https://www.0crat.com/u/botanicus) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by yegor256** [comment]
@ledestin how do you know this code works? Where is the test?

**Comment by ledestin** [comment]
@yegor256 the tests I've written for upgrade manager cover that. They cover a working solution though.

**Comment by yegor256** [comment]
@ledestin that's exactly my point. If all the tests we have now work without your modifications of the code -- why do we need modifications? Everything is fine the way it is, right? If it's not the case, try to reproduce the problem in the test first.

**Comment by ledestin** [comment]
@yegor256 1 out of 3 is working, 2 are skipped. For them to work, the upgrade scripts must be run. I don't see it as good to test the same functionality in another set of tests.

**Comment by yegor256** [comment]
@ledestin unskip the test then? Here is a simple rule, once again: anything you contribute has to be tested. If it's already tested and tests don't fail -- there is no point in contribution.

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @botanicus[/z](https://www.0crat.com/u/botanicus), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#367](https://github.com/zold-io/zold/issues/367) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@botanicus According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only two issues were found during code review. 
Please confirm that you'll try to find at least three major problems while future reviews.

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 point(s) just awarded to @botanicus[/z](https://www.0crat.com/u/botanicus)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request closed without merging**